### DISCUSSION TITLE
Deactivate/delete elements based on material property

### DISCUSSION THREAD

**Discussion by Traiwit** [idaholab/moose#27583] [Q&A General] [answered]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

Hi guys,

just wondering is there a way to deactivate/delete element based on material property?

for example, if porosity of this element = 1, then don't solve/ignore this element in the system (mimic the excavation)

I checked `CoupledVarThresholdElementSubdomainModifier` this is more like re-grouping blocks to me, my simulation is element-based, all material properties are imported via .csv, I kinda don't want to deal with block (element-set) system.

Thank you!
Traiwit

**Comment by GiudGiud** [comment, marked as answer]
Hello

No it hasnt been added. 
You can add a new object, or you can convert a material property to an auxiliary variable with an MaterialRealAux for example then use the current way
`CoupledVarThresholdElementSubdomainModifier` is exactly how we do this currently. You re-assign the elements you do not care to solve on to a block which does not have any kernels / equations defined on

Guillaume

**Reply by Traiwit** [reply]
Hi @GiudGiud, yep make sense 

also, I don't want a new surface around the new block (I already have pre-defined surfaces), if complement_moving_boundary_name isnt specified, will it still create a new surface?

also for the Kernels, is there a way to input all blocks except this new block?

it's an Abaqus mesh, and it has more than a 1000 blocks, which isnt messy when inputting all of them.

Regards,
Traiwit

**Reply by GiudGiud** [reply]
> , if complement_moving_boundary_name isnt specified, will it still create a new surface?

it should not

> also for the Kernels, is there a way to input all blocks except this new block?

I would do a list at the top of the input file that does not include the excluded block
blocks_matter = 'block1 block2 .....' 
then use that for the block restriction

```
[Kernels]
  [k1]
    type = ...
    block = ${blocks_matter}
```

you can use an input file include (with !include file.i) to hide it in another input

**Reply by Traiwit** [reply]
Thank you! this will defs work.

However, I'm just trying to find an easier way to do it (so others in my team can use my input file as well).

is there a way to merge all the subdomains (without having to list all of the subdomains) into 1 subdomain and then apply `CoupledVarThresholdElementSubdomainModifier` to separate the unwanted subdomain out?

or is there a way to print out all the subdomain names?

Regards,
Traiwit

**Reply by GiudGiud** [reply]
> or is there a way to print out all the subdomain names?

You could add that to one of your object? Using `blocks()` from the `BlockRestrictable` interface then printing it with
```
_console << Moose::stringify(blocks()) << std::endl;
```

> is there a way to merge all the subdomains (without having to list all of the subdomains) into 1 subdomain and then apply CoupledVarThresholdElementSubdomainModifier to separate the unwanted subdomain out?

Yes, in a way. You can remove all subdomains but one with a ParsedSubdomainGenerator at the end of the mesh generator step and only leave one.
You can do the same with a `CoupledVarThresholdElementSubdomainModifier` and a well chosen threshold.
But are you trying to remove the blocks from the mesh or not? An element can only have one block.
If you need more flexibility than that, you'll want to work with extra element integer IDs. 
There are less objects for those (than using subdomains/blocks), but the objects missing are easy enough to re-create

**Reply by Traiwit** [reply]
merge all subdomains into 1 wont work, I have different element types in my mesh (c3d6 and c4d6)

back to print all the print out all the subdomain names

I tried to make a userobject that does this, but I got this error during the simulation

```
*** ERROR ***
The parameter "_dual_restrictable" is being retrieved before being set.
```

below is my .C

```
#include "PrintBlockNames.h"
#include "MooseMesh.h"
#include "BlockRestrictable.h"

registerMooseObject("moose_behApp", PrintBlockNames);

InputParameters
PrintBlockNames::validParams()
{
  InputParameters params = GeneralUserObject::validParams();
  params.addClassDescription("Prints the names of the blocks that the object is restricted to.");
  return params;
}

PrintBlockNames::PrintBlockNames(const InputParameters & parameters) :
  GeneralUserObject(parameters),
  BlockRestrictable(this)

{
}

void
PrintBlockNames::execute()
{

    _console << Moose::stringify(blocks()) << std::endl;

}
```

.h

```
#include "GeneralUserObject.h"
#include "BlockRestrictable.h"

class PrintBlockNames : public GeneralUserObject, public BlockRestrictable
{
public:
  static InputParameters validParams();

  PrintBlockNames(const InputParameters & parameters);
  virtual void execute() override;
  virtual void initialize() override {}
  virtual void finalize() override {}

protected:

};
```


more question, is there a way to store it in text, so I can just pass it to the kernel like `block = PrintBlockNames` 

Kind regards,
Traiwit

**Reply by GiudGiud** [reply]
I think general UOs cannot be block-restricted if you get this error. You ll want to go for just UserObject, or maybe ElementUserObject

**Reply by GiudGiud** [reply]
alternatively just set all the missing parameters in validParams():

```
params.set<bool>("_dual_restrictable") = false;
```

**Reply by Traiwit** [reply]
Hi @GiudGiud,

Thank you for your help, at the end, it still doesn't work: `CoupledVarThresholdElementSubdomainModifier` groups both type of element that is above/below the threshold into the same subdomains, and i got this error

```
    Still Outputting Step....libMesh terminating:
Error: Exodus requires all elements with a given subdomain ID to be the same type.
Can't write both PRISM6 and TET4 in the same block!
[0] ../src/mesh/exodusII_io_helper.C, line 2836, compiled Nov 30 2023 at 16:46:29
```

it's okay, this approach is way too try-hard 

I come up with another method, which is just modify the Kernel, I want the Kernel to just ignore when _state[_qp] = 1

is below current? i'm not sure what to return, is 0 the right thing to return when i want to ignore

```
  if (_state[_qp] == 1.0) {
    return 0.0;  // Skip computation for this element
  }
```
```

Real
PorousFlowMassTimeDerivative::computeQpResidual()
{

  if (_state[_qp] == 1.0) {
    return 0.0;  // Skip computation for this element
  }
  Real mass = 0.0;
  Real mass_old = 0.0;
  for (unsigned ph = 0; ph < _num_phases; ++ph)
  {
    const Real dens = (_multiply_by_density ? (*_fluid_density)[_i][ph] : 1.0);
    mass += dens * _fluid_saturation_nodal[_i][ph] * _mass_frac[_i][ph][_fluid_component];
    const Real dens_old = (_multiply_by_density ? (*_fluid_density_old)[_i][ph] : 1.0);
    mass_old +=
        dens_old * _fluid_saturation_nodal_old[_i][ph] * _mass_frac_old[_i][ph][_fluid_component];
  }
  const Real strain = (_has_total_strain ? (*_total_strain_old)[_qp].trace() : 0.0);

  return _test[_i][_qp] * (1.0 + strain) * (_porosity[_i] * mass - _porosity_old[_i] * mass_old) /
         _dt;
}

Real
PorousFlowMassTimeDerivative::computeQpJacobian()
{
  // If the variable is not a PorousFlow variable (very unusual), the diag Jacobian terms are 0
  if (!_var_is_porflow_var)
    return 0.0;
  return computeQpJac(_dictator.porousFlowVariableNum(_var.number()));
}

Real
PorousFlowMassTimeDerivative::computeQpOffDiagJacobian(unsigned int jvar)
{
  if (_state[_qp] == 1.0) {
    return 0.0;  // Skip computation for this element
  }
  // If the variable is not a PorousFlow variable, the OffDiag Jacobian terms are 0
  if (_dictator.notPorousFlowVariable(jvar))
    return 0.0;
  return computeQpJac(_dictator.porousFlowVariableNum(jvar));
}

Real
PorousFlowMassTimeDerivative::computeQpJac(unsigned int pvar)
{
  if (_state[_qp] == 1.0) {
    return 0.0;  // Skip computation for this element
  }
  const unsigned nearest_qp = (_strain_at_nearest_qp ? (*_nearest_qp)[_i] : _i);

  const Real strain = (_has_total_strain ? (*_total_strain_old)[_qp].trace() : 0.0);

  // porosity is dependent on variables that are lumped to the nodes,
  // but it can depend on the gradient
  // of variables, which are NOT lumped to the nodes, hence:
  Real dmass = 0.0;
  for (unsigned ph = 0; ph < _num_phases; ++ph)
  {
    const Real dens = (_multiply_by_density ? (*_fluid_density)[_i][ph] : 1.0);
    dmass += dens * _fluid_saturation_nodal[_i][ph] * _mass_frac[_i][ph][_fluid_component] *
             _dporosity_dgradvar[_i][pvar] * _grad_phi[_j][nearest_qp];
  }

  if (_i != _j)
    return _test[_i][_qp] * (1.0 + strain) * dmass / _dt;

  // As the fluid mass is lumped to the nodes, only non-zero terms are for _i==_j
  for (unsigned ph = 0; ph < _num_phases; ++ph)
  {
    if (_multiply_by_density)
      dmass += (*_dfluid_density_dvar)[_i][ph][pvar] * _fluid_saturation_nodal[_i][ph] *
               _mass_frac[_i][ph][_fluid_component] * _porosity[_i];
    const Real dens = (_multiply_by_density ? (*_fluid_density)[_i][ph] : 1.0);
    dmass += dens * _dfluid_saturation_nodal_dvar[_i][ph][pvar] *
             _mass_frac[_i][ph][_fluid_component] * _porosity[_i];
    dmass += dens * _fluid_saturation_nodal[_i][ph] *
             _dmass_frac_dvar[_i][ph][_fluid_component][pvar] * _porosity[_i];
    dmass += dens * _fluid_saturation_nodal[_i][ph] * _mass_frac[_i][ph][_fluid_component] *
             _dporosity_dvar[_i][pvar];
  }
  return _test[_i][_qp] * (1.0 + strain) * dmass / _dt;
}
```
above is from `PorousFlowMassTimeDerivative.C`

Thank you!
Traiwit

**Reply by GiudGiud** [reply]
A bunch of 0 residuals is likely to create entire rows of 0s in the system, which will cause convergence issues because the solution is no longer unique. 

I think the `CoupledVarThresholdElementSubdomainModifier` ought to be fixed to have an option to not group elements with different element types. @hugary1995 
That would be a fairly straightforward change, just add a new parameter for the alternate group for tets / hexes and assign accordingly

We actually support elements in the same block with different types, except for exodus output.

**Reply by GiudGiud** [reply]
ok looking at it, maybe there should be one `CoupledVarThresholdElementSubdomainModifier` for each type of element. That should already work

Moving elements to inactive subdomains is our main way of doing this. 
There options in PETSc to deal with the null space a bunch of no-kernel regions will create, but I have not ever used this

**Reply by Traiwit** [reply]
i see, so modify Kernel will make convergence worse.

> maybe there should be one CoupledVarThresholdElementSubdomainModifier for each type of element. That should already work

how do i specify each type of element into each `CoupledVarThresholdElementSubdomainModifier`, it doesn't seem to have that option.
doing sub Block inside `CoupledVarThresholdElementSubdomainModifier` is also as difficult, my Abaqus mesh blocks are kinda messy and I have no control over it.

![image](https://github.com/idaholab/moose/assets/79562439/b1b58cbe-d6f0-42c6-bc74-0bd6787004d0)

**Reply by GiudGiud** [reply]
> how do i specify each type of element into each CoupledVarThresholdElementSubdomainModifier, it doesn't seem to have that option.

that would need to be implemented

> doing sub Block inside CoupledVarThresholdElementSubdomainModifier is also as difficult, my Abaqus mesh blocks are kinda messy and I have no control over it.

Looks like the element type is already baked in the name. So that can help us make a list of all the prism blocks and all the tet blocks

Two ways I can see to hack this together right now until @hugary1995 weighs in on how we should do this for the subdomain-modifier:

A more input-file based way:
- print out all the blocks of the mesh in an object (with that line of code I mentioned)
- use a python script or "grep" to get both groups of blocks
- make a second input file with:
```
blocks_tet =  (the output of the script)
blocks_prism = 
```
- add !include "second_input.i" in main input file
- in the main input file, restrict two subdomain modifier UOs with block = ${blocks_tet} and block = ${blocks_prism} respectively

A more programmatic way:
- add a "setBlocks" API to BlockRestrictable.C
- create a parameter in `CoupledVarThresholdElementSubdomainModifier` to pick an element type
- list out all the blocks from that object, use the name for now to pick which one is tet / prism
- setBlocks with one list in one `CoupledVarThresholdElementSubdomainModifier`, setBlocks with the other list in the other one

**Reply by GiudGiud** [reply]
Third way, just use one UserObject and fix the output

either:
- use a new user object that separates blocks by element types.
There is already a mesh generator that does this, inside the MeshRepairGenerator.
This would be a simple ElementUserObject that does this.

- add code to Exodus output in libmesh to split the element types before output
This is actually probably my favorite solution

**Reply by Traiwit** [reply]
Hi @GiudGiud 

I think the first way you mentioned, i still dont get these 2 parts:

> add !include "second_input.i" in main input file
> in the main input file, restrict two subdomain modifier UOs with block = ${blocks_tet} and block = ${blocks_prism} respectively

my round plan would be:

1. list all the block name first (still figuring this out).

2. use `RenameBlockGenerator` to merge all the TET into 1 black and all the Wedge into 1 blocks (MOOSE use BLOCK and Subdomain interchangeably which is a bit confusing, please correct me if i'm wrong)

3. then use CoupledVarThresholdElementSubdomainModifier for each type of BLOCK to separate the inactive block out (this case I will get inactive TET and inactive Wedge blocks as a result).

4. only feed the 2 active blocks (1 TET and 1 Wedge) into Kernel. 

do you think this will work?

Kind regards,
Traiwit

**Reply by GiudGiud** [reply]
I think this will work

You can use the MeshDiagnosticsGenerator or the mesh-info mode (show_info in the mesh generators) to list our the blocks

Block and subdomains are indeed interchangeable

**Reply by GiudGiud** [reply]
Actually to change all the subdomains you could 

Use a parsedSubdomainGenerator to uniformise the entire mesh to a single block

Use the MeshRepairGenerztor to split by element types

**Reply by Traiwit** [reply]
Hi @GiudGiud 

This works like a charm

```
  [subdomains]
    type = ParsedSubdomainMeshGenerator
    input = side_west
    combinatorial_geometry = 'x >= 0 &  y >= 0 & z >= 0'
    block_id = 1
  []

  [separate]
    type = MeshRepairGenerator
    input = 'subdomains'
    separate_blocks_by_element_types = true
  []
```

![image](https://github.com/idaholab/moose/assets/79562439/81a3ad2e-649c-44dd-bc7d-aef82056cf31)

Thank you very much!
Traiwit

**Reply by Traiwit** [reply]
Hi @GiudGiud,

when i use `CoupledVarThresholdElementSubdomainModifier`

```
[UserObjects]
  [1]
    type = CoupledVarThresholdElementSubdomainModifier
    coupled_var = 'state'
    block = 2
    criterion_type = EQUAL
    threshold = 1
    subdomain_id = 4
    execute_on = 'TIMESTEP_BEGIN'
  []
  [2]
    type = CoupledVarThresholdElementSubdomainModifier
    coupled_var = 'state'
    block = 3
    criterion_type = EQUAL
    threshold = 1
    subdomain_id = 5
    execute_on = 'TIMESTEP_BEGIN'
  []
[]
```
it gives this error

```
*** ERROR ***
The following error occurred in the object "mesh", of type "MeshGeneratorMesh".

Requested subdomain 4 does not exist.
```

it only allows me to use subdomain_id = 2 and 3, but If i do that, nothing will happen since it will put the separated elements back into the original subdomain.

I think I need to create an empty (container) subdomains (subdomain_id  4 and 5 in this case), but i'm not sure how to do that?
even better if `CoupledVarThresholdElementSubdomainModifier` can just generate a new subdomain_id  

Do you have any suggestion?

**Reply by GiudGiud** [reply]
Add it to the subdomains in the [Mesh] block to pre-declare it

**Reply by GiudGiud** [reply]
the `add_subdomain_ids` / `add_subdomain_names` parameters

**Reply by Traiwit** [reply]
this works! thank you very much @GiudGiud !

**Reply by Traiwit** [reply]
Hi @GiudGiud 

the simulation does converge after adding this block below to the input file (I haven't even done any active/inactive blocks in Kernels)

```
[UserObjects]
  [1]
    type = CoupledVarThresholdElementSubdomainModifier
    coupled_var = 'state_M0'
    block = 1  # tet element block
    criterion_type = EQUAL
    threshold = 1 
    subdomain_id = 9999 # tet element excav block
    execute_on = 'TIMESTEP_BEGIN TIMESTEP_END'
  []

  [2]
    type = CoupledVarThresholdElementSubdomainModifier
    coupled_var = 'state_M0'
    block = 2  # wedge element block
    criterion_type = EQUAL
    threshold = 1
    subdomain_id = 10000 # wedge  element excav block
    execute_on = 'TIMESTEP_BEGIN TIMESTEP_END'
  []
[]
```

does `CoupledVarThresholdElementSubdomainModifier` change anything in my mesh except for separating excav blocks out from the main blocks? 

with `CoupledVarThresholdElementSubdomainModifier`, at NL step 200+ it still doesnt converge

```
0 Nonlinear |R| = 3.283687e+13
    Computing Jacobian..
      Finished Building Node To Elem Map                                                 [  8.97 s] [52043 MB]
    Still Computing Jacobian.                                                            [ 35.31 s] [52043 MB]

    Computing Residual...                                                                [ 23.02 s] [52105 MB]
    |residual|_2 of individual variables:
            porepressure_L1: 3.21421e+13
 1 Nonlinear |R| = 3.214207e+13
    Computing Jacobian..                                                                 [ 19.59 s] [52105 MB]

    Computing Residual...                                                                [ 23.53 s] [52107 MB]
    |residual|_2 of individual variables:
            porepressure_L1: 3.09547e+13
 2 Nonlinear |R| = 3.095470e+13
    Computing Jacobian..                                                                 [ 19.81 s] [52107 MB]

    Computing Residual...                                                                [ 22.83 s] [52107 MB]
    |residual|_2 of individual variables:
            porepressure_L1: 3.07653e+13
 3 Nonlinear |R| = 3.076527e+13
    Computing Jacobian..                                                                 [ 19.52 s] [52107 MB]

    Computing Residual...                                                                [ 23.25 s] [52107 MB]
    |residual|_2 of individual variables:
            porepressure_L1: 2.97122e+13
 4 Nonlinear |R| = 2.971220e+13
    Computing Jacobian..                                                                 [ 19.63 s] [52107 MB]

    Computing Residual...                                                                [ 23.18 s] [52108 MB]
    |residual|_2 of individual variables:
            porepressure_L1: 2.84524e+13
 5 Nonlinear |R| = 2.845241e+13
    Computing Jacobian..                                                                 [ 19.67 s] [52108 MB]

```
without `CoupledVarThresholdElementSubdomainModifier`,

```
Time Step 1, time = 1.833, dt = 0.083
  Finished Executing Controls                                                            [  9.65 s] [56192 MB]
Still Executing
    Computing Initial Residual
      Computing Residual...                                                              [ 25.16 s] [56602 MB]
    Finished Computing Initial Residual                                                  [ 25.17 s] [56602 MB]
    Computing Residual...                                                                [ 24.97 s] [56602 MB]
    |residual|_2 of individual variables:
            porepressure_L1: 7.40026e+13
 0 Nonlinear |R| = 7.400262e+13
    Computing Jacobian....                                                               [ 27.74 s] [56602 MB]
  Linear solve did not converge due to DIVERGED_ITS iterations 200

    Computing Residual...                                                                [ 25.66 s] [56681 MB]
    |residual|_2 of individual variables:
            porepressure_L1: 4.57095e+13
 1 Nonlinear |R| = 4.570951e+13
    Computing Jacobian...                                                                [ 21.96 s] [56681 MB]
  Linear solve did not converge due to DIVERGED_ITS iterations 200

    Computing Residual...                                                                [ 25.14 s] [56681 MB]
    |residual|_2 of individual variables:
            porepressure_L1: 1.7252e+13
 2 Nonlinear |R| = 1.725200e+13
    Computing Jacobian...                                                                [ 22.50 s] [56681 MB]
  Linear solve did not converge due to DIVERGED_ITS iterations 200

    Computing Residual...                                                                [ 25.43 s] [56682 MB]
    |residual|_2 of individual variables:
            porepressure_L1: 5.22508e+09
 3 Nonlinear |R| = 5.225085e+09
    Computing Jacobian...                                                                [ 22.82 s] [56682 MB]

    Computing Residual...                                                                [ 25.66 s] [56683 MB]
    |residual|_2 of individual variables:
            porepressure_L1: 1.90909e+09
 4 Nonlinear |R| = 1.909095e+09
    Computing Jacobian...                                                                [ 22.33 s] [56683 MB]

    Computing Residual...                                                                [ 25.67 s] [56683 MB]
    |residual|_2 of individual variables:
            porepressure_L1: 1.5773e+09
 5 Nonlinear |R| = 1.577299e+09
    Computing Jacobian...                                                                [ 21.84 s] [56683 MB]

    Computing Residual...                                                                [ 25.46 s] [56684 MB]
    |residual|_2 of individual variables:
            porepressure_L1: 2.96096e+08
 6 Nonlinear |R| = 2.960956e+08
Nonlinear solve converged due to CONVERGED_FNORM_RELATIVE iterations 6
 Solve Converged!
  Finished Solving                                                                       [772.15 s][56684 MB]
```

**Reply by GiudGiud** [reply]
Hello

A few questions,
- you are removing elements from the calculation right?
Should you be moving a boundary at the same time to keep imposing a boundary condition where you are removing?

- What does the variable block look like? Did you make sure to block restrict to block 1 and 2?

- Why  `execute_on = 'TIMESTEP_BEGIN TIMESTEP_END'` ? Do you have something executing in between (use Debug/show_execution_order to check maybe)? Because if not you are executing twice in a row (end of a time step, then beginning of the next one).

- The nonlinear residual even without the UO is not going down that low. Is this as far as you can get it to converge?
- Even without the Uo the linear solve is reporting convergence difficulties, have you tried using a different preconditoner?

**Reply by Traiwit** [reply]
Hi @GiudGiud 

> you are removing elements from the calculation right?
Should you be moving a boundary at the same time to keep imposing a boundary condition where you are removing?

The boundaries are predefined via Abaqus mesh, which exactly wrapping around those elements at each step:

![image](https://github.com/idaholab/moose/assets/79562439/998279ae-3371-4bac-b25b-ec3f64850c92)

![image](https://github.com/idaholab/moose/assets/79562439/0b9de4c3-8a94-4804-8d37-bbf74c77a02d)


> What does the variable block look like? Did you make sure to block restrict to block 1 and 2?

I wasn't aware of this, I will try this and report back:

```
[Variables]
  [porepressure_L1]
    type = MooseVariable
    order = FIRST
    family = LAGRANGE
    block = '1 2'
  []
[]
```

> Why execute_on = 'TIMESTEP_BEGIN TIMESTEP_END' ? Do you have something executing in between (use Debug/show_execution_order to check maybe)? Because if not you are executing twice in a row (end of a time step, then beginning of the next one).

`TIMESTEP_BEGIN` should be enough, since STATE will be updated at `TIMESTEP_BEGIN` as well (read from csv - element based)

> The nonlinear residual even without the UO is not going down that low. Is this as far as you can get it to converge?
> Even without the Uo the linear solve is reporting convergence difficulties, have you tried using a different preconditoner?

yep first step will mostly be taken care by nl_rel_tol, but later it will go as low as 1e4 to 1e5, this is as low as i can push it down, the system is very nonlinear due to (partially saturation and discontinuity). Below is my solver setup

```
[Preconditioning]
  [SMP]
    type = SMP
    full = true
  []
[]

[Executioner]
  type = Transient   # always use 'Transient' (not 'Steady'), see [Kernels]
  solve_type = NEWTON
  petsc_options = '-snes_converged_reason'
  petsc_options_iname = '-pc_type -pc_hypre_type'
  petsc_options_value = 'hypre    boomeramg'
  nl_rel_tol = 1e-5
  nl_abs_tol = 1e5
  nl_max_its = 20
  nl_forced_its = 3
  l_tol = 1e-6
  l_max_its = 200
  # line_search = 'none'
  ###
  ### time stepping:
  # dt = 1.0
  dtmin = 0.083
  dtmax = 1.0
  start_time = 1.75
  end_time = 3
  [TimeStepper]
    ### option-2: define time steps here, directly
    type = TimeSequenceStepper
    time_sequence = '1.000 1.500 1.750 1.833 2.000 2.083 2.167 2.250 2.333 2.417 2.500 . . . . . very long
  []
[]

```

I can also send you the full input file if you'd like, but it's quite long like 5000+ lines

Thank you for you help!

Kind regards,
Traiwit

**Reply by GiudGiud** [reply]
> The boundaries are predefined via Abaqus mesh, which exactly wrapping around those elements at each step:

I m not sure I understand this properly as it seems unlikely that Abaqus knows which elements are getting removed (moved to an inactive block) beforehand, and the boundary at the edge of the removal could not be defined this way?
What is the boundary condition at the boundary near removed (moved to an inactive block) elements?

**Reply by Traiwit** [reply]
Hi @GiudGiud, so we predefine the excavation and drain surface for each timestep in Abaqus mesh. The image below shows the surfaces not the blocks

![image](https://github.com/idaholab/moose/assets/79562439/2454252d-7d4d-418f-b266-c32885c937fd)

then we use multiple BCs + Controls to change the drain surface at each timestep, for example:

```
  [bc_drain0005_UG]
    type = PorousFlowPiecewiseLinearSinkBEH
    variable = porepressure_L1
    boundary = 'DRAINSRFUG_F005_Y2019_M12'
    pt_vals = '0 1e9'
    multipliers = '0 1e9'
    flux_function = 1
    #charElLength = charElLength_M0
    PT_shift = 0.
    fluid_phase = 0
    use_mobility = true    # default is false
    use_relperm = false     # default is false
  []
  
  ### --- TOPO-RECHARGE AND DRAINAGE FOR 0006: F006_Y2020_M01
  ### NOT DEFINING [bc_topol0006_LGFlux]
  ### NOT DEFINING [bc_topol0006_Rainfall]
  ### NOT DEFINING [bc_topol0006_RiverLake]
  ### NOT DEFINING [bc_drain0006_Pit]
  [bc_drain0006_UG]
    type = PorousFlowPiecewiseLinearSinkBEH
    variable = porepressure_L1
    boundary = 'DRAINSRFUG_F006_Y2020_M01'
    pt_vals = '0 1e9'
    multipliers = '0 1e9'
    flux_function = 1
    #charElLength = charElLength_M0
    PT_shift = 0.
    fluid_phase = 0
    use_mobility = true    # default is false
    use_relperm = false     # default is false
  []
  
  ### --- TOPO-RECHARGE AND DRAINAGE FOR 0007: F007_Y2020_M02
  ### NOT DEFINING [bc_topol0007_LGFlux]
  ### NOT DEFINING [bc_topol0007_Rainfall]
  ### NOT DEFINING [bc_topol0007_RiverLake]
  ### NOT DEFINING [bc_drain0007_Pit]
  [bc_drain0007_UG]
    type = PorousFlowPiecewiseLinearSinkBEH
    variable = porepressure_L1
    boundary = 'DRAINSRFUG_F007_Y2020_M02'
    pt_vals = '0 1e9'
    multipliers = '0 1e9'
    flux_function = 1
    #charElLength = charElLength_M0
    PT_shift = 0.
    fluid_phase = 0
    use_mobility = true    # default is false
    use_relperm = false     # default is false
  []

```

```
  [controls0005]
    type = ConditionalFunctionEnableControl
    enable_objects = '    BCs::bc_drain0005_UG'
    # conditional_function = 'if(t>=2.000 & t<2.083,1,0)'
    # execute_on = 'INITIAL TIMESTEP_BEGIN TIMESTEP_END'
    conditional_function = 'if(t>1.833 & t<=2.000,1,0)'
    execute_on = 'INITIAL TIMESTEP_BEGIN'
  []
  [controls0006]
    type = ConditionalFunctionEnableControl
    enable_objects = '    BCs::bc_drain0006_UG'
    # conditional_function = 'if(t>=2.083 & t<2.167,1,0)'
    # execute_on = 'INITIAL TIMESTEP_BEGIN TIMESTEP_END'
    conditional_function = 'if(t>2.000 & t<=2.083,1,0)'
    execute_on = 'INITIAL TIMESTEP_BEGIN'
  []
  [controls0007]
    type = ConditionalFunctionEnableControl
    enable_objects = '    BCs::bc_drain0007_UG'
    # conditional_function = 'if(t>=2.167 & t<2.250,1,0)'
    # execute_on = 'INITIAL TIMESTEP_BEGIN TIMESTEP_END'
    conditional_function = 'if(t>2.083 & t<=2.167,1,0)'
    execute_on = 'INITIAL TIMESTEP_BEGIN'
  []
```

The BC is exactly like `PorousFlowPiecewiseLinearSink` when `charElLength = charElLength_M0` isnt provided.

we want to do inactive elements because we think it might help with convergence + reduce the RAM requirement a bit. 

The simulation is working fine without inactive elements but we are just not sure what material property to assign to those elements to mimic the inactive behaviour.

**Reply by GiudGiud** [reply]
> The simulation is working fine without inactive elements but we are just not sure what material property to assign to those elements to mimic the inactive behaviour.

@cpgr do you know
